crates/ide-completion/src/context/analysis.rs

```diff
@@ -54,226 +59,334 @@ pub(super) fn expand_and_analyze(
     // make the offset point to the start of the original token, as that is what the
     // intermediate offsets calculated in expansion always points to
     let offset = offset - relative_offset;
-    let expansion =
-        expand(sema, original_file, speculative_file, offset, fake_ident_token, relative_offset);
+    let expansion = expand(
+        sema,
+        original_file.clone(),
+        speculative_file.clone(),
+        offset,
+        fake_ident_token.clone(),
+        relative_offset,
+    )
+    .unwrap_or(ExpansionResult {
+        original_file,
+        speculative_file,
+        original_offset: offset,
+        speculative_offset: offset,
+        fake_ident_token,
+        derive_ctx: None,
+    });
 
     // add the relative offset back, so that left_biased finds the proper token
-    let offset = expansion.offset + relative_offset;
+    let offset = expansion.original_offset + relative_offset;
     let token = expansion.original_file.token_at_offset(offset).left_biased()?;
 
     analyze(sema, expansion, original_token, &token).map(|(analysis, expected, qualifier_ctx)| {
-        AnalysisResult { analysis, expected, qualifier_ctx, token, offset }
+        AnalysisResult { analysis, expected, qualifier_ctx, token, original_offset: offset }
     })
 }
 
 /// Expand attributes and macro calls at the current cursor position for both the original file
 /// and fake file repeatedly. As soon as one of the two expansions fail we stop so the original
 /// and speculative states stay in sync.
+///
+/// We do this by recursively expanding all macros and picking the best possible match. We cannot just
+/// choose the first expansion each time because macros can expand to something that does not include
+/// our completion marker, e.g.:
+/// ```
+/// macro_rules! helper { ($v:ident) => {} }
+/// macro_rules! my_macro {
+///     ($v:ident) => {
+///         helper!($v);
+///         $v
+///     };
+/// }
+///
+/// my_macro!(complete_me_here)
+/// ```
+/// If we would expand the first thing we encounter only (which in fact this method used to do), we would
+/// be unable to complete here, because we would be walking directly into the void. So we instead try
+/// *every* possible path.
```

Comment:
Wouldn't it suffice to try paths until we find one where we don't end up in a token tree (where we would lose all semantic info)?

I guess ultimately we would want to pick the one that makes the most sense (as per the ranking, like is done for semantic highlighting and other IDE stuff).

Reply:
Yes, the problem is ranking.

It is possible to optimize this a bit if we give up the *best of all* ranking and chose *the best in the first non-macro expansion*. This will still require looping over the tokens to calculate the offset. Can do that if requested, didn't do that because it seems it won't cause problems to just expand everything.

Reply:
Ye I don't think perf will matter here much, if there is a highly recursive macro somewhere completions will become sluggish either way. So considering all seems the way to go (especially as at some point we'd do https://github.com/rust-lang/rust-analyzer/issues/11058 which requires that as well anyways)